include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h

```diff
@@ -78,9 +102,9 @@
 #    endif
 #endif // _ONEDPL_DETECT_SPIRV_COMPILATION
 
-#if _ONEDPL_LIBSYCL_VERSION >= 50300
+#if _ONEDPL_SYCL2020_REQD_SUB_GROUP_SIZE_PRESENT
 #    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) sycl::reqd_sub_group_size(SIZE)
-#else
+#elif _ONEDPL_LIBSYCL_VERSION_LESS_THAN(50300)
```

Comment:
What if both macros evaluated to zero?
1) If it is possible, we don't control the all cases and get UB/error...
2) If it is not possible, I don't understand why do we write the second if  -  `#elif`...

Reply:
Of course we do not control all the cases, because we cannot predict what an arbitrary SYCL implementation can or cannot do, and what workaround it would require.

The optimistic expectation is that fully SYCL-conformant implementations go into the "present" branch with no effort from our side. However if for some implementation that's not true, the case (1) will happen and we will need to analyze the failure and provide a proper workaround, likely in another #elif branch.

Reply:
Specifically for `_ONEDPL_SYCL_REQD_SUB_GROUP_SIZE`, we can define it as no-op in the #else branch. But this might result in runtime errors for the kernels that really require certain WG size, which would be even worse than the guaranteed compilation error.

Reply:
Another approach is to switch from "present" macros to "workaround" macros. I think Dmitry tried something of that kind with "absent" macros in a previous revision of the patch. It would be something like this:

```
#if _ONEDPL_LIBSYCL_REQD_SUB_GROUP_SIZE_WA // == _ONEDPL_LIBSYCL_VERSION_LESS_THAN(50300)
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) intel::reqd_sub_group_size(SIZE)
#else // SYCL 2020 conformance expected
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) sycl::reqd_sub_group_size(SIZE)
#endif
```

For an arbitrary implementation of SYCL it is no better than the proposed way, because if `sycl::reqd_sub_group_size` is not supported, there will also be a compilation error. And while the code appears to be slightly simpler, a downside is that the conformant code comes last, not first.

Reply:
Let me add more details on top of Alexey's answer.

> What if both macros are evaluated to zero... If it is possible

It is possible.

Let's consider such a case: older versions of FOOSYCL, some SYCL library, do not support `sycl::reqd_sub_group_size`. There is no workaround, or it was simply forgotten to be added. 

```c++
#define _ONEDPL_SYCL2020_REQD_SUB_GROUP_SIZE_PRESENT \
    (!_ONEDPL_LIBSYCL_VERSION_LESS_THAN(50300) && !_ONEDPL_FOOSYCL_VERSION_LESS_THAN(42))

// Current approach: older versions of FOOSYCL will have undefined macro, which may have unpredictable consequences
#if _ONEDPL_SYCL2020_REQD_SUB_GROUP_SIZE_PRESENT
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) sycl::reqd_sub_group_size(SIZE)
#elif _ONEDPL_LIBSYCL_VERSION_LESS_THAN(50300)
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) intel::reqd_sub_group_size(SIZE)
#endif

// Possible approach: error will be emitted with the older FOOSYCL versions, but it is more verbose
#if _ONEDPL_SYCL2020_REQD_SUB_GROUP_SIZE_PRESENT
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) sycl::reqd_sub_group_size(SIZE)
#elif _ONEDPL_LIBSYCL_VERSION_LESS_THAN(50300)
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) intel::reqd_sub_group_size(SIZE)
#else
#    error "No implementation for reqd_sub_group_size"
#endif
```

Was it your concern, Mikhail? If so, would you prefer "Possible approach" above?

Reply:
> we can define it as no-op in the #else branch

We can also define `#pragma error` in the #else branch... 
I think compile time error is better  then runtime error.

Reply:
> if `sycl::reqd_sub_group_size` is not supported, there will also be a compilation error

Ok, this behavior is acceptable for me.

Reply:
> ```c++
> #    error "No implementation for reqd_sub_group_size"
> ```

Yes, this approach is acceptable to me

Reply:
> Was it your concern, Mikhail? If so, would you prefer "Possible approach" above? It's an edge-case, I am reluctant to do that across the PR due to the verbosity.

Sorry, but it seems I didn't catch what you want to say...

Reply:
> We can also define #pragma error in the #else branch...
I think compile time error is better then runtime error.

Done.

> It's an edge-case, I am reluctant to do that across the PR due to the verbosity.

It was about proliferation of "#else... #error ..." parts, which I assumed as non-necessary. It is not relevant, I changed my mind: it is better to have these parts to avoid UB in some cases.